[V4] Query engine (#31)
strapi/rfcs: rfcs/xxx-v4-query-engine.md

- Start Date: 2021-07-01
- RFC PR: (leave this empty)

# Summary

> In v4 we are build a new database layer to support the futur evolution of Strapi and we are introducing a new query engine with it.

Reviewer: ```suggestion
In v4 we are build a new database layer to support the future evolution of Strapi and we are introducing a new query engine along with it.
```

The query engine should mostly be used by plugin developers, developers adding custom business logic to their applications and the Strapi core team.

# Example

```js

> qb.query("article").findMany({

Reviewer: ```suggestion
db.query("article").findMany({
```

  where: {
    title: {
      $startWith: "2021",
      $endsWith: "v4",
    },
  },
  populate: {
    category: true,
  },
});
```

# Motivation

> We want to make sure we have a more flexible and powerfull database layer to build the futur features of Strapi.

Reviewer: ```suggestion
We want to make sure we have a more flexible and powerful database layer to build future features of Strapi onto.
```

Moreover we want to provide a better experience and more possibilities for plugin developers and application developers to build advanced business logic into their plugins or applications.

> This layer is really made for database queries and is at a lower level than our previous version. This layer isn't abstracting a lot of things so we can have full control.

Reviewer: ```suggestion
This layer is really made for database queries and is at more low-level than our previous version. This layer isn't abstracting a lot of things so developers can have full control.
```

# Detailed design

## Query API

> To work with entities in the database you can user the Query api:

Reviewer: ```suggestion
To work with entities in the database you can use the Query api:
```

  select: ["title", "description"],
  where: { title: "Hello World" },
  orderBy: { title: "DESC" },
  populate: { category: true },
});
```

### `findMany(params)` ⇒ `Entry[]`

Finds entries matching the params

**Params**

| Param    | Type                           | Description                                |
| -------- | ------------------------------ | ------------------------------------------ |
| select   | `string[]`                     | Select some attributes to return           |
| where    | [`WhereParam`](#filtering)     | Filter                                     |
| limit    | `number`                       | How many entries to return                 |
| offset   | `number`                       | How many entries to skip                   |
| orderBy  | [`OrderByParam`](#ordering)    | How to order the entries                   |
| populate | [`PopulateParam`](#populating) | Select which relations should be populated |

**Example**

```js
const entries = await db.query("article").findMany({
  select: ["title", "description"],
  where: { title: "Hello World" },
  orderBy: { title: "DESC" },
  populate: { category: true },
});
```

> ### `findWithCount(params)` => `[Entry[], number]`

Reviewer: Should this method be called `findManyWithCount` to be more consistent?

  data: {
    title: "My Article",
  },
});
```

### `update(params)` => `Entry`

Updates one entry and returns it

**Params**

| Param    | Type                           | Description                                |
| -------- | ------------------------------ | ------------------------------------------ |
| select   | `string[]`                     | Select some attributes to return           |
| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
| where    | [`WhereParam`](#filtering)     | Filter                                     |
| data     | `object`                       | Input data                                 |

**Example**

```js
const entry = await db.query("article").update({
  where: { id: 1 },
  data: {
    title: "xxx",
  },
});
```

### `delete(params)` => `Entry`

Deletes one entry and returns it

**Params**

| Param    | Type                           | Description                                |
| -------- | ------------------------------ | ------------------------------------------ |
| select   | `string[]`                     | Select some attributes to return           |

> | populate | [`PopulateParam`](#populating) | Select which relations should be populated |

Reviewer: Does `populate` make sense for delete queries? I'm not sure what I'd expect from that ... does it control the `cascade` behavoir?

Author: This has the same behaviour create & update to be consistent

| where | [`WhereParam`](#filtering) | Filter              |
| data  | `object`                   | Array of input data |

**Example**

```js
await db.query("article").updateMany({
  where: {
    price: 20,
  },
  data: {
    price: 18,
  },
});

// { count: 42 }
```

### `deleteMany(params)` => `{ count: number }`

Deletes multiple entries matching the params

**Params**

| Param | Type                       | Description |
| ----- | -------------------------- | ----------- |
| where | [`WhereParam`](#filtering) | Filter      |

**Example**

```js
await db.query("article").deleteMany({
  where: {
    title: {
      $startsWith: "v3",
    },
  },
});

// { count: 42 }
```

## Aggregations

### `count(params)` => `number`

Counts entries matching the params

**Params**

| Param | Type                       | Description |
| ----- | -------------------------- | ----------- |
| where | [`WhereParam`](#filtering) | Filter      |

```js
const count = await db.query("article").count({
  where: {
    title: {
      $startsWith: "v3",
    },
  },
});

// 12
```

### Avg,Min,Max,Sum

> The rest of this API (avg,min,max,sum,groupby) will specified at a later time.

## Query Builder

> When in need fore more control over the queries than with the Query API you can use the underlying query builder.

Reviewer: ```suggestion
When you are in need for more control over the queries the Query API generates, you can use the underlying query builder.
```

**Example**

```js
const entries = db.query("article").findMany({
  where: {
    title: {
      $startsWith: "ABCD",
    },
  },
});
```

#### `$endsWith`

Attribute ends with input value.

**Example**

```js
const entries = db.query("article").findMany({
  where: {
    title: {
      $endsWith: "ABCD",
    },
  },
});
```

#### `$null`

Attribute is null`.

**Example**

```js
const entries = db.query("article").findMany({
  where: {
    title: {
      $null: true,
    },
  },
});
```

#### `$notNull`

Attribute is not null.

**Example**

```js
const entries = db.query("article").findMany({
  where: {
    title: {
      $notNull: true,
    },
  },
});
```

## Populating

With this new database layer, relations and components have a unified API for populating them.

You can populate by passing an array of attribute names

```js
db.query("article").findMany({
  populate: ["componentA", "relationA"],
});
```

For more advanced usage you can pass an object

```js
db.query("article").findMany({
  populate: {
    componentB: true,
    dynamiczoneA: true,
    relation: someLogic || true,
  },
});
```

You can also apply where filters and select or populate nested relations

```js

> db.query("article").findMany({

Author: NOTE: Extends the example a bit more
